### DISCUSSION TITLE
Web server failed: listen tcp 192.168.1.126:59999: bind: cannot assign requested address

### DISCUSSION THREAD

**Discussion by Liguangwei0321** [edgexfoundry/community#220] [Q&A]
when run device_simple, get errors:
<img width="1078" alt="error" src="https://github.com/edgexfoundry/community/assets/45187340/4161aca0-4782-4cae-86f1-7467c5071697">

level=ERROR ts=2023-11-30T03:01:41.415677205Z app=device-simple source=httpserver.go:143 msg="Web server failed: listen tcp 192.168.1.126:59999: bind: cannot assign requested address

**Comment by jiekechoo** [comment]
You could try run app using`root` .
```
sudo ./device-simple
```
Here is a guide you can refer:
https://docs.edgexfoundry.org/3.1/getting-started/Ch-GettingStartedHybrid/

**Reply by Liguangwei0321** [reply]
i try use "sudo ./device-simple",but it display : level=WARN ts=2023-11-30T07:18:58.141047832Z app=device-simple source=secret.go:111 msg="Retryable failure while creating SecretClient: open /tmp/edgex/secrets/device-simple/secrets-token.json: no such file or directory"

**Reply by jiekechoo** [reply]
@Liguangwei0321 
Run `export EDGEX_SECURITY_SECRET_STORE=false` in terminal before start the device service. You should run it in `insecure` mode.
here is a example:
```
sudo su
export EDGEX_SECURITY_SECRET_STORE=false
./device-simple -cp -d -o
```

**Reply by bnevis-i** [reply]
> You could try run app using root .

One shouldn't need root to bind any port > 1024.    Some paths in /tmp/edgex/secrets/device-simple might need to be re-acled in order to allow the service to read its secret store token outside of docker.  Most likely, the bind failure is because there is a process already listening on the port.  After this is fixed, I suggest going back and finding out exactly what privilege is neeed.  Root processes listening on open ports is hazardous.

**Reply by lenny-goodell** [reply]
@Liguangwei0321 , this is not an issue with the port, but the IP address. You can not use the host system IP in docker container since that IP doesn't exist in the container. You must use `localhost` and then expose the port to the host system via `ports` in the compose file.

**Comment by eaton-coreymutter** [comment]
If this is Linux and not containerized: you can see if any process is already listening on that port:

```
sh-5.0# fuser 59880/tcp
516
sh-5.0# ps | grep 516
  516 nobody   1296m S    /usr/bin/core-data
```

If there isn't anything already listening on that port, I have seen this problem when some other process is using that port as a _client_. For example, some other service has a connection to core-metadata on port 59881, and got randomly assigned 59999 as the source port, keeping you from using that port as a destination. 
To fix that, we do `echo 40000-60000 > /proc/sys/net/ipv4/ip_local_reserved_ports` as root early in bootup, preventing those ports from getting randomly assigned as source ports.

**Reply by lenny-goodell** [reply]
Per the error message `bind: cannot assign requested address` it is an issue with the IP address, not the port.
